### ISSUE TITLE
Property not set in constructor when using trait in abstract class

### ISSUE THREAD

**Issue by gianiaz** [vimeo/psalm#6087]
I think I've found a strange behaviour, in this snippet psalm complains about a property not set in constructor when the concrete class extends an abstract class using a trait.

https://psalm.dev/r/5a9fe1453e

If I take rid of the abstract class everything is fine.

https://psalm.dev/r/13cce84db7

It's a bug?

**Comment by Jean85** [comment]
Issue is still present event with an explicit call to `parent::__construct(...)`: https://psalm.dev/r/f177e0b295

**Comment by Jean85** [comment]
It seems related to #4393

**Comment by weirdan** [comment]
The issue is valid because descendants can override `initTimestamp()` (https://3v4l.org/gOaFu). But I think it should be reported on `Foo` rather than on `Bar`. Also, marking `initTimestamp()` final should prevent it, but it doesn't seem to happen currently: https://psalm.dev/r/1041e80bc6

**Comment by Jean85** [comment]
You probably would need to make the constructor final too, but the issue is still present there too: https://psalm.dev/r/e0c6c4b20f

Maybe the analysis should change approach and determine if the call sequence in any class does end with the initialization of such properties..

**Comment by rickogden** [comment]
I've also been having this issue when attempting to write a Psalm plugin for Doctrine ODM for MongoDB. It's slightly different, as the constructor itself is in the trait, but the error is the same.

Example: https://psalm.dev/r/0f71ca2564

**Comment by rulatir** [comment]
> The issue is valid because descendants can override `initTimestamp()` (https://3v4l.org/gOaFu). But I think it should be reported on `Foo` rather than on `Bar`. Also, marking `initTimestamp()` final should prevent it, but it doesn't seem to happen currently: https://psalm.dev/r/1041e80bc6

No, the issue is not valid. There is no *can* in "descendants can override". A descendant either *does* initialize the `creationDate` property in its constructor by some means, or it *does not*. If it does, all is fine. If it doesn't (because it overrides `initTimestamp()`, or inherits an overriden `initTimestamp()` from a descendant-in-the-middle, or fails to call parent constructor) then the problem is with this very descendant that doesn't. 100% of the information needed to determine which is the case is available to Psalm statically.

**Comment by jaikdean** [comment]
Are there any additional test cases or anything else that would help with this issue? I'd like to contribute but have next to no understanding of how Psalm's internals work.

**Comment by orklah** [comment]
Mainly we need contributors with a good understanding of how Traits works. I barely use them myself so I'm not very motivated to dig into this subject on top of being almost ignorant about the syntax and how people use them.

If you want to take a shot, you can try following those instructions: https://github.com/vimeo/psalm/issues/6937#issuecomment-972847445 and report your findings, we'll try to help if you get stuck
